### ISSUE TITLE
write_hdf5 with sector stream reader slows significantly

### ISSUE THREAD

**Issue by ercius** [OpenChemistry/stempy#128]
Im having trouble with the write_hdf5() function for 1kx1k scans with FileVersion.VERSION4 (sector) files.  The initial speed is very fast but it eventually saturates and takes extremely long times to complete. I tracked the output H5 file size as a function of time:
![image](https://user-images.githubusercontent.com/15657108/74563117-e8b5d580-4f20-11ea-91a9-2fb84788f755.png)

I let it run about 6 hours once and only got about 500 GB of the full 760GB and the drive was writing extremely slowly.

I could imagine a few reasons why this might happen. It might be that the RAM is filled up, the drives overheat and slow down, or the file gets so large it gets too slow to write to.

I think the last possibility is unlikely since the H5 file seems to have the correct chunking and the fall off in speed occurs at a specific time.

Here is the RAM usage from `top` at the end of the chart:
```
top - 10:58:56 up 8 days, 20:01,  2 users,  load average: 1.09, 1.13, 1.43
Tasks: 357 total,   2 running, 222 sleeping,   0 stopped,   0 zombie
%Cpu(s):  3.5 us,  1.5 sy,  0.0 ni, 95.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem : 26362440+total,  1780808 free, 51157912 used, 21068568+buff/cache
KiB Swap:        0 total,        0 free,        0 used. 21016640+avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 2184 daquser   20   0 48.602g 0.047t  24468 R  99.7 19.0  35:30.81 python3
```

I think the process is using 48 GB of RAM. Is that right? Its not the maximum of the system, but very large.

The RAW sector files are on 4 NVME drives and the H5 file is being written to a single spinning HDD. I dont have enough disk space to save the H5 to the NVME to test if its the HDD.

Any ideas on what is going on?

**Comment by cjh1** [comment]
@ercius Interesting! So it looks like we are get as similar data rate up until about 300 GB, at which point it starts to drop off.

I would expect the overhead associated with seeking to the correct place in the file to write a frame to be constant for the whole processes of writing of the file.

What is the percentage of data captured? How make raw sector files are their?

**Comment by PhilippPelz** [comment]
Hmm, just looking through the code, h5readwrite.cpp opens the h5 dataset, selects hyperslab and closes it again for every single frame that is written to hard disk.
Maybe a bottleneck.

L610 reader.cpp
L235 h5readwrite.cpp : H5ReadWrite::updateData

**Comment by psavery** [comment]
The strange thing is that the write rate seems to be fairly constant over time until it hits around 300 GB. It makes me wonder if there's an overflow (or something similar) somewhere.

**Comment by psavery** [comment]
@ercius Are you using the default format (`SectorReader.H5Format.Frame`), or the `DataCube` format?

**Comment by ercius** [comment]
I'm using the default format. There is nothing obvious as to why it would
slow down.

**Comment by cjh1** [comment]
@ercius  It would be good to know the % of data captured. As you know we read all the files in the round robin fashion, reconstructing the frames into a cache, once a full frame is complete its written to the file and its removed from the cache. So one possible explanation for the slowing in the rate a which the file is growing is that the frame aren't be reconstructed ( so no frames or few frames are being written to the cache ) and partial frames are building up in the cache, which could explain the memory usage. Is it possible you are getting data loss towards the end of the scan? We can build in some instrumentation to test this theory, if you don't mind building and running a test branch.

**Comment by ercius** [comment]
@cjh1 Good question and maybe this points to the problem. The data set Im trying to reconstruct is a 1025x1024 scan (659GB in total). It has a total # sectors = 4193847 and percent saved = 99.89%. So, we are missing a relatively small but not objectively insignificant amount of data (~715 MB) depending on the cache size. (FYI this was due to an issue which is now fixed. We get 99.99% now.)

Also interesting is what part of the scan is missing data.  Here is a "map" of the data validity showing # of sectors saved for each scan position (integer values of 0 - 4 are possible):
![Figure_1](https://user-images.githubusercontent.com/15657108/74670888-e9d44600-515e-11ea-86c3-6f4e9ff83dc4.png)
We have a lot of missing sectors in the beginning before the receivers catch up. Then around the middle one receiver drops out for a couple of lines. Maybe the cache fills up just around the middle (which is about 300 GB) and slows the process down?

**Comment by cjh1** [comment]
@ercius So with that loss ratio the cache would only grow to about 2.8 GB ( for every lost sector we allocation a full frame in the cache ). However, lost sectors is not the only way the cache might grown large, if the sectors associated with a frame got widely dispersed in the files then the cache would grown ( as in the sectors associated with a frame did not appear in relatively small number of round robin reads ), but would eventually shrink. Did the write operation ever run to completion?

**Comment by cjh1** [comment]
@psavery Has added some instrumentation that may help here, can we get you to repeat with the instrumented code? Are you using singularity for this?

**Comment by ercius** [comment]
I was not using the singularity container before. I just pulled the most recent one and backed up the other one.

Im going to try to rebuild the data. It will take some time to see the performance. Ill get back to you when I have some results.

**Comment by cjh1** [comment]
Ok, let us know if you want to try running the instrumented branch.

**Comment by ercius** [comment]
Oh. I thought I was running the new branch. The same thing happens with the
singularity image. Right at 300 GB the speed drops.

Let's try the instrumented branch.

**Comment by cjh1** [comment]
@ercius No, that was the image with `imageio`, I will push a docker image for the branch.

**Comment by cjh1** [comment]
Here is the command to pull the instrumented image. It should write debug info to stdout every minute, so if you could redirect it to a file that would be good. 

`singularity pull docker://openchemistry/stempy:debug`

**Comment by psavery** [comment]
@ercius It prints out some information to stdout that we hope will be useful. Make sure you capture the stdout.

**Comment by ercius** [comment]
I am just able to get to this now. I left the write_hdf5() function running with the non-debug dicker image. Here is the timing result:
![chart](https://user-images.githubusercontent.com/15657108/74693527-042f1380-51a1-11ea-93b7-d06bf22e6ef0.png)
Its an inflection point in perfromance. Very odd.

Ill start running the debug image now.

**Comment by ercius** [comment]
Here are the timing results from the debug run
![chart (1)](https://user-images.githubusercontent.com/15657108/74701319-39485f80-51bb-11ea-8cd2-6491549287c1.png)

And here is the stdout
[write_hdf5_debug.txt](https://github.com/OpenChemistry/stempy/files/4217179/write_hdf5_debug.txt)

It initially completes 255 frames per second. Then it suddenly drops to 17 fps and continues to slow down to about 13 fps. The frame cache does not change at the inflection point. In fact it goes down in when the rate drops initially.

So, Im not sure whats going on here. It might be a local issue either with the hardware or the data set. I should test this on another data set and try writing to a different HDD before we continue debugging stempy. Let me know if you have other ideas though.

**Comment by cjh1** [comment]
@ercius As the cache doesn't seem to be growing, it does suggest that the problem is on the writer side. To confirm this we could build another image that just throws the data away, if it a writer problem we should get a pretty constant frame rate. Trying on another dataset and HDD are also good tests todo.

**Comment by ercius** [comment]
I wrote a quick script to mimic the writing of the data. Here is the code:
```python
import numpy as np
import h5py

with h5py.File('/mnt/hdd1/temp.h5','w') as f0:
    d = f0['/'].create_dataset('data',shape=(1,576,576),dtype='<u2',maxshape=(None,576,576))
    aa = np.ones((576,576),dtype='<u2')
    for ii in range(0,600*1024):
        d[ii,:,:] = aa
        d.resize(ii + 2,axis=0)
```
I only wrote enough fake frames to get above the 300 GB threshold. The file increases up to 408 GB with 600*1024 frames. It finished successfully. So, there is no problem with writing a large > 300 GB data set using h5py and hopefully also HDF5 in C++.

Here is the timing chart. I think I missed the end of the file writing so the last point might be later than the actual finish time. Still, I think this proves the point.
![chart](https://user-images.githubusercontent.com/15657108/74791412-3b6bf600-526f-11ea-92cf-a0f3dae54e06.png)

@cjh1 Does this match how the writer in stempy works? I think we should try disabling the writing and see if that allows the process to finish without the slowdown.

**Comment by cjh1** [comment]
@ercius Yes in general that is what stempy is doing, although its hard to know for sure without knowing how that various h5py options map to the C API. I will go ahead and update the debug image to disable the writing.

**Comment by cjh1** [comment]
@ercius I have pushed an update image to the 'debug' tag, so the following command should pull it down: 

```
singularity pull docker://openchemistry/stempy:debug
```
Note, it still creates the file, I just commented out any updates.

**Comment by ercius** [comment]
Just tested whether sorting the filenames has an effect. Im using glob to get the list of files, and I thought maybe they were not sorted correctly. The same behavior is seen though when I use sorted() on the `list` of files.

**Comment by cjh1** [comment]
@ercius Did you manage to test with the latest debug image?

**Comment by ercius** [comment]
@cjh1 I just pulled it and started running the code. We are getting about 10x faster frame building  (~2000 fps) without writing. Ill let you know if  it slows down.

**Comment by ercius** [comment]
It finished without any slowdown. Here is the output
[stempy-debug-no-writing.txt](https://github.com/OpenChemistry/stempy/files/4237645/stempy-debug-no-writing.txt)

**Comment by cjh1** [comment]
@ercius Thank you running that test. We will do some digging into the HDF5 written to see if we can see what the problem is there.

**Comment by cjh1** [comment]
@ercius I have been able to create a test case that recreates a similar performance profile using dummy data. So I should be able to investigate more easily what is going on without needed the dataset.

**Comment by cjh1** [comment]
@ercius Ok, I think we have got to the bottom of this, I have pushed image with a fix, could you give it a try when you get a chance?

```
singularity pull docker://openchemistry/stempy:fix
```

Its related to the way get where getting the dataset dimensions, it seems to be more costly than simple metadata, so we are now just passing it in.

**Comment by ercius** [comment]
@cjh1 I just started running this. Ill have an answer in the afternoon when I finish up a few meetings. Thanks for tracking this down! What a weird problem though.

**Comment by cjh1** [comment]
@ercius Yes, very strange, we had assumed it was just static meta data, apparently not!

**Comment by ercius** [comment]
@cjh1 it worked! It took about 53 minutes to process a full 1025x1024 data set. Thanks for figuring that out. The fix can be merged into the master. Let me know when I can pull a new image.

@PhilippPelz  It should be on NERSC now. ncemhub/2020.01.13/data_scan143.h5

**Comment by PhilippPelz** [comment]
Awesome! Thank you all!
Peter, what is the file name/scan number?

Peter Ercius <notifications@github.com> schrieb am Mi., 26. Feb. 2020,
17:23:

**Comment by ercius** [comment]
@PhilippPelz ncemhub/2020.01.13/data_scan143.h5

**Comment by cjh1** [comment]
That's great news, we will get the fix merged today.

**Comment by cjh1** [comment]
This fix is now merge into master. The latest image can be pulled using: 

```
singularity pull docker://openchemistry/stempy:latest
```

I have also updated the shifter image.
